### ISSUE TITLE
Last layer elision

### ISSUE THREAD

**Issue by mil-ad** [KaidiXu/auto_LiRPA#5]
Do you do last layer elision similar to the original IBP? I saw some references to `C_merged`, is it doing something similar?

if yes it'd be great if you could describe that bit of the code a little bit. I'm a bit confused about the shape of tensors as this merging happens. I think `C` should have have the shape `batch * (num_classes - 1) * num_classes)` and IBP kind of merges all specs into one step by constructing the worst-case-logits (picking lower bound for correct class and upper bound for _all_ the wrong classes). But I'm not sure how this changes when we do this merging step.

Keep up the good work.

**Comment by mil-ad** [comment]
btw I'm particularly talking about the forward propagation and _not_ crown style backward style.

**Comment by KaidiXu** [comment]
Hi,

Thanks for your question.
Yes, we did a similar way as IBP to merge the C matrix with the last linear layer (without loss fusion). You can refer to the code start from here:
https://github.com/KaidiXu/auto_LiRPA/blob/7e1fbf12d857ef8d411d80eef1bd73d9ae4ba3be/auto_LiRPA/bound_general.py#L727
Then we merge the C matrix here:
https://github.com/KaidiXu/auto_LiRPA/blob/7e1fbf12d857ef8d411d80eef1bd73d9ae4ba3be/auto_LiRPA/bound_ops.py#L686
the shapes are C: [batch, (num_classes - 1), num_classes)], w: [num_classes, num_neurons], so the merged w would have shape: [batch, (num_classes - 1), num_neurons].
Finally, it multiplies with the previous h_L and h_U at here: https://github.com/KaidiXu/auto_LiRPA/blob/7e1fbf12d857ef8d411d80eef1bd73d9ae4ba3be/auto_LiRPA/bound_ops.py#L700
The shape of h_L and h_U is [batch, num_neurons], so the output shape will change to [batch, (num_classes - 1)] at:https://github.com/KaidiXu/auto_LiRPA/blob/7e1fbf12d857ef8d411d80eef1bd73d9ae4ba3be/auto_LiRPA/bound_ops.py#L653 which implies the worst-case-logits as you said.

Best,
Kaidi Xu
________________________________

**Comment by mil-ad** [comment]
Thanks for the quick reply Kaidi, this is getting clearer to me. One more questions what are `lb` and `ub` and how are they different from the actual bounds? And why are they being multiplied by C when merging? What are their shapes?

```py
if C is not None:
    w = C.matmul(w)
    lb = C.matmul(lb) if not isinstance(lb, float) else lb
    ub = C.matmul(ub) if not isinstance(ub, float) else ub
```

**Comment by KaidiXu** [comment]
They are lower/upper bias of the linear layer. If no perturbation on the bias, then they are the same value actually with the shape [output_neurons]. Refer to the IBP paper https://arxiv.org/pdf/1810.12715.pdf eq. (9).

Best,
Kaidi Xu
________________________________

**Comment by mil-ad** [comment]
Ah of course! Do we typically have perturbation on the bias?

**Comment by KaidiXu** [comment]
No, most existing works only perturb on input.

Best,
Kaidi Xu
________________________________

**Comment by huanzhang12** [comment]
In verification setting we typically do not have perturbation on model weights and bias. But our framework does support perturbation on parameter space (model weight or bias), as they are also inputs of the computational graph. In [our paper](https://arxiv.org/pdf/2002.12920.pdf) we demonstrated that training a model with weight perturbations can lead to smooth optimization landscape and better generalization.
